uc-crux-llvm/src/UCCrux/LLVM/Classify/Poison.hs

```diff
@@ -0,0 +1,171 @@
+         What4.asConcrete op1,
+         getTermAnn sym annotations op2,
+         What4.asConcrete op2
+       ) of
+    (Just ann, Nothing, Nothing, Just val) ->
+      Just (ann, What4.fromConcreteBV val)
+    (Nothing, Just val, Just ann, _) ->
+      Just (ann, What4.fromConcreteBV val)
+    _ -> Nothing
+
+handleBVOp ::
+  (MonadIO f, What4.IsExprBuilder sym) =>
+  AppContext ->
+  FunctionContext m arch argTypes ->
+  sym ->
+  -- | Term annotations (origins)
+  Map (Some (What4.SymAnnotation sym)) (Some (TypedSelector m arch argTypes)) ->
+  MissingPreconditionTag ->
+  What4.SymBV sym w ->
+  What4.SymBV sym w ->
+  (forall w'. 1 <= w' => NatRepr w' -> BV w' -> Constraint m ('FTInt w')) ->
+  f (Maybe (Explanation m arch argTypes))
+handleBVOp appCtx funCtx sym annotations tag op1 op2 constraint =
+  case annotationAndValue sym annotations op1 op2 of
+    Just (Some (TypedSelector ftRepr (SomeInSelector (SelectArgument idx cursor))), concreteSummand) ->
+      do
+        let argName i =
+              funCtx ^. argumentNames . ixF' i . to getConst . to (maybe "<top>" Text.unpack)
+        liftIO $
+          (appCtx ^. log) Hi $
+            Text.unwords
+              [ "Diagnosis:",
+                diagnose tag,
+                "#" <> Text.pack (show (Ctx.indexVal idx)),
+                "at",
+                Text.pack (show (ppCursor (argName idx) cursor))
+              ]
+        liftIO $ (appCtx ^. log) Hi (prescribe tag)
+        case ftRepr of
+          FTIntRepr w ->
+            pure $
+              Just $
+                ExMissingPreconditions
+                  ( tag,
+                    [ NewConstraint
+                        (SomeInSelector (SelectArgument idx cursor))
+                        (constraint w (BV.trunc' w concreteSummand))
+                    ]
+                  )
+          _ -> pure Nothing
+    _ -> pure Nothing
+
+classifyPoison ::
+  forall f sym m arch argTypes.
+  (MonadIO f, What4.IsExprBuilder sym) =>
+  AppContext ->
+  FunctionContext m arch argTypes ->
+  sym ->
+  -- | Term annotations (origins)
+  Map (Some (What4.SymAnnotation sym)) (Some (TypedSelector m arch argTypes)) ->
+  Poison.Poison (Crucible.RegValue' sym) ->
+  f (Maybe (Explanation m arch argTypes))
+classifyPoison appCtx funCtx sym annotations =
+  \case
+    Poison.AddNoSignedWrap (Crucible.RV op1) (Crucible.RV op2) ->
+      handleBVOp
+        appCtx
+        funCtx
+        sym
+        annotations
+        ArgAddSignedWrap
+        op1
+        op2
+        ( \w concreteSummand ->
+            BVCmp L.Islt w (BV.sub w (BV.maxSigned w) concreteSummand)
+        )
+    Poison.SubNoSignedWrap (Crucible.RV op1) (Crucible.RV op2) ->
+      handleBVOp
+        appCtx
+        funCtx
+        sym
+        annotations
+        ArgSubSignedWrap
+        op1
+        op2
+        ( \w concreteSummand ->
+            BVCmp L.Isgt w (BV.add w (BV.minSigned w) concreteSummand)
+        )
+    _ -> pure Nothing
```

Comment:
Running this tool on a larger program uncovered a bug here: This code assumes that `concreteSummand` is positive, otherwise the constraints generated can be nonsensical and cause `Simulate.hs` to add a concretely false constraint (e.g. that a given signed integer is less than the most negative signed integer of that size). Shouldn't be too hard to fix, I'll do so tomorrow.

Reply:
I *may* have been able to catch this in the test suite if I had additional tests like those generated by `oneArith`, except:

- I also had tests for the symmetric version (i.e. `argument - constant` *and* `constant - argument`)
- I tested negative numbers in addition to positive ones

I'll add those tests while/before resolving this.

Reply:
Bug fixed, many tests added, and re-tested on the original program (`libi18n`) on which I saw the panic and it was fixed.